sysdeps/nacl/ifaddrs.c

```diff
@@ -17,811 +17,144 @@
-		    case IFA_CACHEINFO:
-		      break;
-		    default:
-		      break;
-		    }
-
-		  rta = RTA_NEXT (rta, rtasize);
-		}
+	char* buf = malloc(IFADDRS_BUFSIZE);
+	int result = __nacl_irt_getifaddrs(buf, IFADDRS_BUFSIZE);
 
-	      /* If we didn't get the interface name with the
-		 address, use the name from the interface entry.  */
-	      if (ifas[ifa_index].ifa.ifa_name == NULL)
-		ifas[ifa_index].ifa.ifa_name
-		  = ifas[map_newlink (ifam->ifa_index - 1, ifas,
-				      map_newlink_data, newlink)].ifa.ifa_name;
-
-	      /* Calculate the netmask.  */
-	      if (ifas[ifa_index].ifa.ifa_addr
-		  && ifas[ifa_index].ifa.ifa_addr->sa_family != AF_UNSPEC
-		  && ifas[ifa_index].ifa.ifa_addr->sa_family != AF_PACKET)
-		{
-		  uint32_t max_prefixlen = 0;
-		  char *cp = NULL;
-
-		  ifas[ifa_index].ifa.ifa_netmask
-		    = &ifas[ifa_index].netmask.sa;
-
-		  switch (ifas[ifa_index].ifa.ifa_addr->sa_family)
-		    {
-		    case AF_INET:
-		      cp = (char *) &ifas[ifa_index].netmask.s4.sin_addr;
-		      max_prefixlen = 32;
-		      break;
-
-		    case AF_INET6:
-		      cp = (char *) &ifas[ifa_index].netmask.s6.sin6_addr;
-		      max_prefixlen = 128;
-		      break;
-		    }
-
-		  ifas[ifa_index].ifa.ifa_netmask->sa_family
-		    = ifas[ifa_index].ifa.ifa_addr->sa_family;
-
-		  if (cp != NULL)
-		    {
-		      char c;
-		      unsigned int preflen;
-
-		      if ((max_prefixlen > 0) &&
-			  (ifam->ifa_prefixlen > max_prefixlen))
-			preflen = max_prefixlen;
-		      else
-			preflen = ifam->ifa_prefixlen;
-
-		      for (i = 0; i < (preflen / 8); i++)
-			*cp++ = 0xff;
-		      c = 0xff;
-		      c <<= (8 - (preflen % 8));
-		      *cp = c;
-		    }
-		}
-	    }
+	if (result < 0) {
+		free(buf);
+		errno = result;
+		return -1;
 	}
-    }
 
-  assert (ifa_data_ptr <= (char *) &ifas[newlink + newaddr] + ifa_data_size);
+	const char s[2] = "\n";
+	char *token;
+	struct ifaddrs *ifa;
 
-  if (newaddr_idx > 0)
-    {
-      for (i = 0; i < newlink; ++i)
-	if (map_newlink_data[i] == -1)
-	  {
-	    /* We have fewer links then we anticipated.  Adjust the
-	       forward pointer to the first address entry.  */
-	    ifas[i - 1].ifa.ifa_next = &ifas[newlink].ifa;
-	  }
+	ifa = malloc(sizeof(struct ifaddrs));
+	*ifap = ifa;
 
-      if (i == 0 && newlink > 0)
-	/* No valid link, but we allocated memory.  We have to
-	   populate the first entry.  */
-	memmove (ifas, &ifas[newlink], sizeof (struct ifaddrs_storage));
-    }
+	/* get the first token */
+	token = strtok(buf, s);
 
-  *ifap = &ifas[0].ifa;
+	/* walk through other tokens */
+	while( token != NULL ) {
 
- exit_free:
-  __netlink_free_handle (&nh);
-  __netlink_close (&nh);
+		ifa->ifa_name = calloc(1, sizeof(char) * IF_NAMESIZE);
+		ifa->ifa_addr = malloc(sizeof(struct sockaddr));
+		ifa->ifa_netmask = malloc(sizeof(struct sockaddr));
+		ifa->ifa_broadaddr = malloc(sizeof(struct sockaddr));
+		int flags;
 
-  return result;
+		char name[IF_NAMESIZE];
+		char addr[16] = {0};
```

Comment:
You will also need more than 16 bytes if you were planning on expanding to IPv6.